curve-amm/src/lib.rs

```diff
@@ -0,0 +1,1062 @@
+            } else if k != j {
+                x_k = *xp_k;
+            } else {
+                continue;
+            }
+            // s = s + x_k
+            s = s.checked_add(&x_k)?;
+            // c = c * d / (x_k * n)
+            c = c.checked_mul(&d_f)?.checked_div(&x_k.checked_mul(&n)?)?;
+        }
+        // c = c * d / (ann * n)
+        // At this step we have d^n in the numerator of c
+        // and n^(n-1) in its denominator.
+        // So we multiplying it by remaining d/n
+        c = c.checked_mul(&d_f)?.checked_div(&ann_f.checked_mul(&n)?)?;
+
+        // b = s + d / ann
+        // We subtract d later
+        let b = s.checked_add(&d_f.checked_div(&ann_f)?)?;
+        let mut y = d_f;
+
+        for _ in 0..255 {
+            let y_prev = y;
+            // y = (y^2 + c) / (2 * y + b - d)
+            // Subtract d to calculate b finally
+            y = y
+                .checked_mul(&y)?
+                .checked_add(&c)?
+                .checked_div(&two.checked_mul(&y)?.checked_add(&b)?.checked_sub(&d_f)?)?;
+
+            // Equality with the specified precision
+            if y > y_prev {
+                if y.checked_sub(&y_prev)? <= self.get_precision() {
+                    return Some(y);
+                }
+            } else if y_prev.checked_sub(&y)? <= self.get_precision() {
+                return Some(y);
+            }
+        }
+
+        None
+    }
+
+    fn create_pool(
+        &mut self,
+        who: &ActorId,
+        assets: Vec<ActorId>,
+        pool_asset: &ActorId,
+        amplification_coefficient: FixedU128,
+        fee: Permill,
+        admin_fee: Permill,
+    ) -> Option<PoolId> {
+        // Assets related checks
+        if assets.len() < 2 {
+            panic!("create_pool : please provide atleast two assets");
+        }
+        let unique_assets = BTreeSet::<ActorId>::from_iter(assets.iter().copied());
+        if unique_assets.len() != assets.len() {
+            panic!("create_pool : duplicate assets found");
+        }
+
+        // Add new pool
+        let pool_id = self.pool_count;
+
+        // We expect that PoolInfos have sequential keys.
+        // No PoolInfo can have key greater or equal to PoolCount
+        if self.pools.get(&pool_id).is_some() {
+            panic!("create_pool : inconsistent storage");
+        }
+
+        let pool_info = PoolInfo {
+            owner: *who,
+            pool_asset: *pool_asset,
+            assets,
+            amplification_coefficient,
+            fee,
+            admin_fee,
+        };
+        self.pools.insert(pool_id, pool_info);
+
+        self.pool_count = pool_id.checked_add(1).expect("inconsistent storage");
+
+        Some(pool_id)
+    }
+
+    fn fixed_to_u128(value: &FixedU128) -> u128 {
+        value.into_inner() / FixedU128::DIV
+    }
+
+    pub async fn get_pool_balances(assets: &[ActorId]) -> Vec<FixedU128> {
+        let mut balances = Vec::new();
+        for asset in assets {
```

Comment:
Isn't `select_all` more preferable in such kind of processing? Refer to an example:

https://github.com/gear-tech/gear/blob/master/examples/async-multisig/src/lib.rs#L79